### ISSUE TITLE
Images as attachments?

### ISSUE THREAD

**Issue by vloexo** [mhoeher/opentodolist#46]
Hello again :)

I'm not sure if this is an expected use case at all, but I would like to use the attachment feature of the app(s) to sync (some) images - relevant to the corresponding note - across my device. For example, say you have a note called trees and want to include a picture of a birch tree in the note.

The problem is, that this does not seem to work. While I can choose images as attachments fine, upon synchronization I will get an error message saying 

`"Synchronization failed for library XXXX: Uploading /2021/04/xxxx.jpg" failed: Connection closed.".`

Although I'm not entirely sure the problem is not on my end, I can safely say that for example syncing images using Nextcloud normally works fine for my setup.

Any advice on this will be greatly appreciated :)

**Comment by mhoeher** [comment]
Hm... generally, attaching files and syncing them should work without issues. I also just tested here locally, with v3.33 of the app. Attaching a file and synchronizing it works flawlessly. I am also using NextCloud.

Could you maybe provide some additional information on your used setup?

- Where do you run the app (mobile or Desktop?)
- Is the problem reproducible? When you attach additional images (or other files), are they properly uploaded or is there also an error shown?
- How large are the files you try to upload?
- If you host NextCloud on your own: Could you maybe also check the server logs? Is something visible there? From the log message, it seems the HTTP connection to the server could be established successfully, but the server unexpectedly closed it - so a server side log probably would be the most helpful thing here.
- If running on a desktop system, you could also try to start the app from the command line to collect additional debug output as a last resort. However, we probably would have to increase verbosity so that logs of the underlying Qt framework are included - however, I don't know if they will be too useful in this case. As per the previous point, my suspicion is that the server closes the connection for whatever reason (maybe due to the app does the request in an invalid way). So server side logs would probably be the more helpful ones here to understand what is going wrong.

**Comment by vloexo** [comment]
Thanks for your answer.

First of all, here's a rough summary of my setup:
- Desktop:
Linux Mint 20.1 running Cinnamon 4.8.6
OpenTodoList App Version: 3.33.0 (installed via Snap)

- Phone:
Moto Z Play running unofficial Lineage OS 16.0 (Android 9 Pie)
OpenTodoList App Version: 3.33.0 (installed via F-Droid)

- Nextcloud:
Version: 20.0.7
Linux version: 3.2.40 armv7l (Synology DSM)
PHP version: 7.3.16
mysql version: 10.3.24 (MariaDB)

I just found out, that the general syncing of attachments actually does work indeed, this just has to be done file-by-file. I was trying to sync multiple images at once before, which might have been too much for my little NAS setup.
So with single files I have tried uploading PDF and image files as attachments from the desktop app which worked fine.

One thing I found, trying to upload files as attachments from my phone is the following: In the phone app, when inside the respective note, upon selecting the "attach file" entry, a file selection dialogue pops up. The thing is that this dialogue starts in some (app-?)folder, with folders `/app_outdex, /cache, /code_cache, /files, /lib` and `/qt-reserved-files`. Inside the `/files` folder, there is another `/cache` directory as well as multiple `/{xxxx}` folders, which I guess are the note libraries or something similar.
Now here's the thing: When I try to go up a directory level, i.e. leaving the app folder, I am presented a blank screen, and am thus unable to select any files to upload as attachments.
Note that this only happens with the android version of the app, and only in the file selection dialogue for attachment uploads. Uploading an image as distinct note, for example, works as intended.

Another thing I just realized is the following: After successful uploading and syncing of attachments, I can not look at them. What I mean by this is, they are listed in the respective note but I can can not click on them or select them in another way to look at the content of the attachment. Deleting them works fine though.

One last thing I have found while playing around is a bug in the picture upload functionality of the desktop app. When in the main view of a library, there is a little plus in the lower left corner. When I select that plus and the choose the "picture" or "image" entry, the whole app just dies. This is also reproducible on my system.
(Note: Pardon me if the actual strings/names of entries are different, I am not using the software in English.)

Thank you for time, if any more information is needed from my side, I will be happy to try and provide it.

**Comment by mhoeher** [comment]
> I just found out, that the general syncing of attachments actually does work indeed, this just has to be done file-by-file. I was trying to sync multiple images at once before, which might have been too much for my little NAS setup.
> So with single files I have tried uploading PDF and image files as attachments from the desktop app which worked fine.

Well, this is good and bad news, I'd say :wink: So for you, you can probably continue to work, but (for me, as the app developer) I am still interested in the root cause :sunglasses: Did you have a chance to look into the server side logs? Was anything visible there concerning that sync issue?

One "guess" I would make is this: The app currently uses up to twelve connections in parallel to download and upload data. This is to use bandwidth more efficiently - OpenTodoList uses a lot of usually very small files. Synchronizing them one by one would be utterly slow. I could, however, imagine, that if the server side has to deal with several files that are being uploaded simultaneously where the files potentially are "large" (whatever large means...), it might cancel certain transactions to not exhaust resources. But again, this is just guesswork and without logs this will be hard to prove.

> One thing I found, trying to upload files as attachments from my phone is the following: In the phone app, when inside the respective note, upon selecting the "attach file" entry, a file selection dialogue pops up. The thing is that this dialogue starts in some (app-?)folder, with folders `/app_outdex, /cache, /code_cache, /files, /lib` and `/qt-reserved-files`. Inside the `/files` folder, there is another `/cache` directory as well as multiple `/{xxxx}` folders, which I guess are the note libraries or something similar.
> Now here's the thing: When I try to go up a directory level, i.e. leaving the app folder, I am presented a blank screen, and am thus unable to select any files to upload as attachments.
> Note that this only happens with the android version of the app, and only in the file selection dialogue for attachment uploads. Uploading an image as distinct note, for example, works as intended.

Well, let's say it like that: The current file selector (on Android and probably also iOS) is rather sub-optimal :wink: Somewhere one one of my many todo lists, I think I still have a todo that says "use proper file picker for generic files on Android". I think up to now the attachment feature is not so widely used - I personally also use it rather seldom - but as we already have a proper image picker in place, doing the same for generic files on Android shouldn't be too hard to get done, so I'd maybe be able to add this in the June release. I created an issue for this over at GitLab, here is the [link](https://gitlab.com/rpdev/opentodolist/-/issues/438).

> Another thing I just realized is the following: After successful uploading and syncing of attachments, I can not look at them. What I mean by this is, they are listed in the respective note but I can can not click on them or select them in another way to look at the content of the attachment. Deleting them works fine though.

Well, yes, another weak point. Currently, to "view" attachments, I use a built-in function of the underlying Qt framework. This function basically asks the operating system to open the specified file. The point is: On Android, the attachments (as the complete library) are stored in the app's private folder, so this call definitely will fail on Android (and also iOS).

I created [another issue over at Gitlab](https://gitlab.com/rpdev/opentodolist/-/issues/439) to look into this. It's also scheduled for now for the June release.

> One last thing I have found while playing around is a bug in the picture upload functionality of the desktop app. When in the main view of a library, there is a little plus in the lower left corner. When I select that plus and the choose the "picture" or "image" entry, the whole app just dies. This is also reproducible on my system.
> (Note: Pardon me if the actual strings/names of entries are different, I am not using the software in English.)

Erm... okay. That's interesting. I just tested with the snap version of the app (I am using Fedora Linux) - there, adding images works as expected without any problems. Could you start the app from the command line and check if something meaningful is printed to the console?

**Comment by vloexo** [comment]
> Well, this is good and bad news, I'd say wink So for you, you can probably continue to work, but (for me, as the app developer) I am still interested in the root cause sunglasses Did you have a chance to look into the server side logs? Was anything visible there concerning that sync issue?

Of course :) When trying to sync multiple images, the server (sooner or later) fails and logs the following:

`[no app in context] Error: Sabre\DAV\Exception\BadRequest: Expected filesize of 1921351 bytes but read (from Nextcloud client) and wrote (to Nextcloud storage) 1114112 bytes. Could either be a network problem on the sending side or a problem writing to the storage on the server side. at <<closure>>`

`0. /volume1/web/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php line 156
   OCA\DAV\Connector\Sabre\File->put(null)`
`1. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 1104
   OCA\DAV\Connector\Sabre\Directory->createFile("image.png", null)`
`2. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 527
   Sabre\DAV\Server->createFile("OpenTodoList/{8 ... g", null, null)`
`3. /volume1/web/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php line 89
   Sabre\DAV\CorePlugin->httpPut(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})`
`4. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 474
   Sabre\DAV\Server->emit("method:PUT", [Sabre\HTTP\Requ ... }])`
`5. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 251
   Sabre\DAV\Server->invokeMethod(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})`
`6. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 319
   Sabre\DAV\Server->start()`
`7. /volume1/web/nextcloud/apps/dav/appinfo/v1/webdav.php line 84
   Sabre\DAV\Server->exec()`
`8. /volume1/web/nextcloud/remote.php line 167
   require_once("/volume1/web/ne ... p")`

`PUT *server url of file, IP, user, timestamp*`

Which is then followed by:
`[PHP] Error: Error: Cannot modify header information - headers already sent by (output started at /volume1/web/nextcloud/3rdparty/sabre/http/lib/Sapi.php:132) at /volume1/web/nextcloud/apps/dav/lib/Connector/Sabre/File.php#691 at <<closure>>`

`0. <<closure>>
   OC\Log\ErrorHandler::onError(2, "Cannot modify h ... )", "/volume1/web/ne ... p", 691, {string: "X-Hash ... "})`
`1. /volume1/web/nextcloud/apps/dav/lib/Connector/Sabre/File.php line 691
   header("X-Hash-SHA256:  ... c")`
`2. /volume1/web/nextcloud/apps/dav/lib/Connector/Sabre/File.php line 195
   OCA\DAV\Connector\Sabre\File->header("X-Hash-SHA256:  ... c")`
`3. <<closure>>
   OCA\DAV\Connector\Sabre\File->OCA\DAV\Connector\Sabre\{closure}("*** sensitive parameters replaced ***")`
`4. /volume1/web/nextcloud/lib/private/Files/Stream/HashWrapper.php line 71
   call_user_func(Closure {}, "*** sensitive parameter replaced ***")`
`5. /volume1/web/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 319
   OC\Files\Stream\HashWrapper->stream_close()`
`6. /volume1/web/nextcloud/apps/dav/appinfo/v1/webdav.php line 84
   Sabre\DAV\Server->exec()`
`7. /volume1/web/nextcloud/remote.php line 167
   require_once("/volume1/web/ne ... p")`

`PUT *server url of file, IP, user, timestamp*`

> Well, let's say it like that: The current file selector (on Android and probably also iOS) is rather sub-optimal wink Somewhere one one of my many todo lists, I think I still have a todo that says "use proper file picker for generic files on Android". I think up to now the attachment feature is not so widely used - I personally also use it rather seldom - but as we already have a proper image picker in place, doing the same for generic files on Android shouldn't be too hard to get done, so I'd maybe be able to add this in the June release. I created an issue for this over at GitLab, here is the [link](https://gitlab.com/rpdev/opentodolist/-/issues/438).

Ah, I see. I just wanted to mention it, but good to see this is a known issue.

> Well, yes, another weak point. Currently, to "view" attachments, I use a built-in function of the underlying Qt framework. This function basically asks the operating system to open the specified file. The point is: On Android, the attachments (as the complete library) are stored in the app's private folder, so this call definitely will fail on Android (and also iOS).

Just to make sure: (for me) This is not just a problem with android, but I can't seem to open attachments using the desktop app on Linux either.

> I created [another issue over at Gitlab](https://gitlab.com/rpdev/opentodolist/-/issues/439) to look into this. It's also scheduled for now for the June release.

Thank you, very much appreciated :)
 
> Erm... okay. That's interesting. I just tested with the snap version of the app (I am using Fedora Linux) - there, adding images works as expected without any problems. Could you start the app from the command line and check if something meaningful is printed to the console?

Ok, I just started the app with the debugging parameter and got this output on the console after trying to add an image:
`(OpenTodoList:135137): GLib-GIO-WARNING **: 10:18:26.960: Error creating IO channel for /proc/self/mountinfo: Permission denied (g-file-error-quark, 2)`

`(OpenTodoList:135137): Gtk-WARNING **: 10:18:27.022: Could not find the icon 'user-home-symbolic-ltr'. The 'hicolor' theme
was not found either, perhaps you need to install it.
You can get a copy from:
	http://icon-theme.freedesktop.org/releases`

I did a quick check on this, and hicolor-icon-theme is installed on my system in version 0.17-2.

`(OpenTodoList:135137): Gtk-WARNING **: 10:18:27.024: Could not load a pixbuf from /org/gtk/libgtk/icons/16x16/status/image-missing.png.
This may indicate that pixbuf loaders or the mime database could not be found.
**`
`Gtk:ERROR:../../../../gtk/gtkiconhelper.c:494:ensure_surface_for_gicon: assertion failed (error == NULL): Failed to load /org/gtk/libgtk/icons/16x16/status/image-missing.png: Unrecognized image file format (gdk-pixbuf-error-quark, 3)
Bail out! Gtk:ERROR:../../../../gtk/gtkiconhelper.c:494:ensure_surface_for_gicon: assertion failed (error == NULL): Failed to load /org/gtk/libgtk/icons/16x16/status/image-missing.png: Unrecognized image file format (gdk-pixbuf-error-quark, 3)
/snap/opentodolist/233/usr/bin/OpenTodoList-launcher: line 21: 135137 Aborted                 (core dumped) $SNAP/usr/bin/OpenTodoList "$@"`

I hope this helps :)

**Comment by mhoeher** [comment]
Thanks for the logs! I just did a quick search and found something that looks suspiciously close to what you describe: https://help.nextcloud.com/t/problem-uploading-files-to-nc18-expected-filesize-of-1024000-bytes-but-read-from-nextcloud-client-and-wrote-to-nextcloud-storage-696024-bytes/69836/13

i.e.:

- Single file uploads usually work just fine, but
- When uploading several images (aka larger files) at once, that error triggers.

So for the time being I would consider this a bug on the server side - not sure if NextCloud or one of the "standard" components in between that are involved. I am afraid, there is nothing on the client side which could prevent this. Well, except that the app should retry uploading the file upon the next sync run. Is this the case? You would then probably still get some errors but eventually the app should finish uploading everything.

Regarding opening the attachments: For me it works smoothly. Could you run the app from the terminal and check if something gets printed to it as soon as you try to open one of the attachments?

And finally, about the crash: Thanks for the logs, but unfortunately, they don't contain what I was hoping for :disappointed: 

I think what is reported is coming from the GTK theme plugin delivered as part of the Qt installation inside the snap. And probably, the icon theme is not available inside the snap - that's why I suspect these errors are shown. But they shouldn't cause the app to malfunction. Could you do me another favour and:

1. Run the app, trying to select an image.
2. Once the crash happened, run the following command: `coredumpctl dump`. This should print some information about the most recent crash (which should then be the app). The dump should also contain a stack trace, which might yield a first hint on what the issue could be.

Thanks!

**Comment by vloexo** [comment]
> So for the time being I would consider this a bug on the server side - not sure if NextCloud or one of the "standard" components in between that are involved. I am afraid, there is nothing on the client side which could prevent this. Well, except that the app should retry uploading the file upon the next sync run. Is this the case? You would then probably still get some errors but eventually the app should finish uploading everything.

Hmm. I don't know if this is considered good practice, but I guess as a workaround it should work, no?

> Regarding opening the attachments: For me it works smoothly. Could you run the app from the terminal and check if something gets printed to it as soon as you try to open one of the attachments?

Unfortunately, nothing is printed to the terminal. Notably, this is also true for other "pure-GUI" actions, that don't trigger synchronization or change the state of the application. Maybe there's is the possibility to somehow get more debugging output? Right now I run the application using the following command:
`opentodolist --qmljsdebugger port:1234`

> The dump should also contain a stack trace, which might yield a first hint on what the issue could be.

Alright, here you go:
[dump.txt](https://github.com/mhoeher/opentodolist/files/6349842/dump.txt)

> Thanks!

No, thank you :)

**Comment by mhoeher** [comment]
> Hmm. I don't know if this is considered good practice, but I guess as a workaround it should work, no?

The thing is I currently see no other reliable way to workaround this. In theory, the client could maybe try to detect exactly this issue - broken upload due to server issues - and immediately retry the upload. But I'm not sure if this could cause more harm than just living with the intermediate errors and wait for the sync to eventually upload everything, especially as other user's might not be working with NextCloud as their backend and there a failed upload could be due to a completely different issue. I think, this is a typical case of "choosing between plague and cholera"; none of the solutions is ideal and no matter what we do, there's no way to make everyone 100% happy :disappointed: 

The important thing is: Are the files eventually uploaded in your case? Or, once the upload for a particular attachment failed, is the file not uploaded upon the next sync?

> > Regarding opening the attachments: For me it works smoothly. Could you run the app from the terminal and check if something gets printed to it as soon as you try to open one of the attachments?
> 
> Unfortunately, nothing is printed to the terminal. Notably, this is also true for other "pure-GUI" actions, that don't trigger synchronization or change the state of the application. Maybe there's is the possibility to somehow get more debugging output? Right now I run the application using the following command:
> `opentodolist --qmljsdebugger port:1234`

What you could try is to run the app like this:

```bash
QT_LOGGING_RULES='*.*=true' opentodolist | 2>&1 | tee ~/opentodolist.log
```

This will generate *a lot* of logs, collecting them in the file `opentodolist.log` in your home directory. You could either try to check the log on your own or - if you trust me enough - provide it to me (ideally, drop me a mail, as I don't know if it might not contain some information that are rather private).

> > The dump should also contain a stack trace, which might yield a first hint on what the issue could be.
> 
> Alright, here you go:
> [dump.txt](https://github.com/mhoeher/opentodolist/files/6349842/dump.txt)

Thanks! I'll dig into this this evening to find out more.

**Comment by vloexo** [comment]
> none of the solutions is ideal and no matter what we do, there's no way to make everyone 100% happy disappointed

Don't get me wrong, I do agree with you, that this is kind of the only way to deal with similar problems on the client side of things. Another thing I want to add here: I am really unsure how common this problem is at all for other users, maybe this is kind of specific to my server setup.

> The important thing is: Are the files eventually uploaded in your case? Or, once the upload for a particular attachment failed, is the file not uploaded upon the next sync?

Well, unfortunately, for now it looks like once it failed, it will always fail. Although this is a little bit hard to definitely test "per hand". But for the last 10 or 15 tries, the same image file that had failed to sync before, would fail to sync again when trying.

> What you could try is to run the app like this:
> 
> ```shell
> QT_LOGGING_RULES='*.*=true' opentodolist | 2>&1 | tee ~/opentodolist.log
> ```

Hmm, this does indeed create some output on the console, but once the application is started, the console task returns and the log file itself thus stays of course empty. For reference, here's the output I get from the console while startup (although I'm not to sure if it's useful):
`QT_LOGGING_RULES='*.*=true' opentodolist | 2>&1 | tee ~/opentodolist.log`
`Entered main`
`Starting app`
`Applying app-wide configs`
`Creating Qt app`
`qt.qpa.xcb: Has MIT-SHM     : true`
`qt.qpa.xcb: Has MIT-SHM FD  : true`
`qt.qpa.xcb: Using XInput version 2.2`
`qt.qpa.screen: Output eDP-1 is not enabled`
`qt.qpa.screen: Output DP-1 is not connected`
`qt.qpa.screen: Output HDMI-1 is not connected`
`qt.qpa.screen: Output DP-2 is not connected`
`qt.qpa.screen: EDID data for output "HDMI-2": identifier '', manufacturer 'Samsung Electric Company',model 'S27E390-', serial '808728889', physical size: 600.00x340.00`
`qt.qpa.xcb: XmbTextListToTextProperty result code -2`
`qt.qpa.screen: adding QXcbScreen(0x560d81be1f50, name="HDMI-2", geometry=1920x1080+0+0, availableGeometry=1920x1035+0+45, devicePixelRatio=1.0, logicalDpi=QPair(106.0,106.0), physicalSize=598.0x336.0mm, screenNumber=0, virtualSize=1920x1080 (1920.0x1080.0mm), orientation=Qt::LandscapeOrientation, depth=24, refreshRate=60.0, root=7a7, windowManagerName="Mutter (Muffin)") (Primary: true )` 
`qt.qpa.screen: primary output is "HDMI-2"`
`qt.qpa.input.devices: input device  Virtual core XTEST pointer ID 4`
`qt.qpa.input.devices:    has 10 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`
`qt.qpa.input.devices: input device  SynPS/2 Synaptics TouchPad ID 11`
`qt.qpa.input.devices:    has 7 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Horiz Scroll" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Vert Scroll" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`
`qt.qpa.input.devices: input device  TPPS/2 IBM TrackPoint ID 12`
`qt.qpa.input.devices:    has 7 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Horiz Scroll" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Vert Scroll" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`
`qt.qpa.input.devices: input device  USB Keyboard Consumer Control ID 17`
`qt.qpa.input.devices:    has 7 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Horiz Scroll" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Vert Scroll" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`
`qt.qpa.input.devices: input device  USB Laser Game Mouse Consumer Control ID 19`
`qt.qpa.input.devices:    has 7 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Horiz Scroll" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Vert Scroll" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`
`qt.qpa.input.devices: input device  USB Laser Game Mouse ID 22`
`qt.qpa.input.devices:    has 20 buttons`
`qt.qpa.input.devices:    has valuator "Rel X" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Y" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Horiz Scroll" recognized? true`
`qt.qpa.input.devices:    has valuator "Rel Vert Scroll" recognized? true`
`qt.qpa.input.devices:    it's a scrolling device`

`(process:327442): Gtk-WARNING **: 13:20:41.575: Locale not supported by C library.
	Using the fallback 'C' locale.`
`Gtk-Message: 13:20:41.575: Failed to load module "xapp-gtk3-module"`
`qt.qpa.xcb: using glib dispatcher`
`qt.xkb.compose: using xkb compose input context`
`Qt: Session management error: Could not open network socket`
`qt.qpa.fonts: default fonts: system 0x560d81d402d0 fixed 0x560d81d402f0`
`Screens:`
`"HDMI-2" 1`
`Parsing command line arguments`
`Processing command line ("/snap/opentodolist/233/usr/bin/OpenTodoList")`
`Finished parsing command line`
`Opening console`
`Checking for primary app instance...`

> > > The dump should also contain a stack trace, which might yield a first hint on what the issue could be.
> > 
> > 
> > Alright, here you go:
> > [dump.txt](https://github.com/mhoeher/opentodolist/files/6349842/dump.txt)
> 
> Thanks! I'll dig into this this evening to find out more.

You're welcome.

**Comment by mhoeher** [comment]
Regarding the sync: Just to make sure I understand you correctly... upon each subsequent sync, the app will try to upload again, but every time the upload fails? That sounds rather strange :open_mouth: I mean: Without attaching the same file again - just swipe down or use "Sync now" from the menu?

Regarding the log: I think the app still runs in the background. You have to terminate it (on Linux, this is done by pressing `Ctrl+Q`). Afterwards, when starting from the terminal, the app should not detach.

**Comment by vloexo** [comment]
> Regarding the sync: Just to make sure I understand you correctly... upon each subsequent sync, the app will try to upload again, but every time the upload fails? That sounds rather strange open_mouth I mean: Without attaching the same file again - just swipe down or use "Sync now" from the menu?

Yes, using the "Sync now" button from the menu. I just tried again, after fetching the log and multiple times opening and closing the app, and the problem still persists.

> Regarding the log: I think the app still runs in the background. You have to terminate it (on Linux, this is done by pressing `Ctrl+Q`). Afterwards, when starting from the terminal, the app should not detach.

Correct, my bad. Also, the actual command that worked for me was the following (mind the missing piping symbol before merging the std streams):
```shell
QT_LOGGING_RULES='*.*=true' opentodolist 2>&1 | tee > ~/opentodolist.log
```
I wanted to send you an email with the log attached, but wasn't able to find one either on here, GitLab or the website you linked on your account.

**Comment by mhoeher** [comment]
:raised_eyebrow: 

Sounds wrong... so the app retries the upload every time, but the server errors out as well. Interesting. Could you check the server logs again? Are, for example, the number of bytes it received and has written it reports always the same or do they differ?

Regarding the log: You can drop me a message via "martin AT rpdev DOT net".

**Comment by mhoeher** [comment]
Regarding the opening of attachments: I found something!

After unsuccessfully searching the log you sent (thanks for that!) I tried to reproduce by switching to the snap version and well, I saw the same issue. Most interestingly, nothing in the overall control flow involved seems to trigger a debug print telling about the issue. Luckily, we're not the only ones suffering from this. I found this forum post:

https://forum.snapcraft.io/t/xdg-open-or-gvfs-open-qdesktopservices-openurl-file-somelocation-file-txt-wont-open-the-file/16824/12

Which exactly sounds like our problem. One of the Snap devs obviously found a bug in Qt regarding opening of documents from within snaps: https://bugreports.qt.io/browse/QTBUG-83939

That one, however, is still open.

If I understand the discussion in the forum correctly, it seems we could workaround this issue by manually using `xdg-open` or `xdg-open-portal` to open files from within the snap.

I created another issue over on Gitlab - see https://gitlab.com/rpdev/opentodolist/-/issues/440

As this actually might be quite easy to implement, I tentatively scheduled for the coming release (due beginning of May).

**Comment by mhoeher** [comment]
Another short update: I just pushed a new version of the app to the snap beta channel. If you like, you can switch to that by running

```bash
sudo snap switch opentodolist --channel edge
sudo snap refresh
```
With this version, at least for me locally everything works as expected :wink:

**Comment by vloexo** [comment]
> Regarding the opening of attachments: I found something!
> ...
> With this version, at least for me locally everything works as expected wink

I just tried the new version and - TADA - it works :) Thank you very much!
Does this fix have anything to do with the similar problem on the android version?

Regarding the file syncing:
> Sounds wrong... so the app retries the upload every time, but the server errors out as well. Interesting. Could you check the server logs again? Are, for example, the number of bytes it received and has written it reports always the same or do they differ?

Hmm so I just checked, and it seems like for every image file - if it fails to upload (getting a `Sabre\DAV\Exception\BadRequest`) - there's three times the error for header modification (see above) and then nothing follows, until the next time the images is (tried to be) synchronized. Interestingly, when triggering synchronization using the "Sync now" entry from the menu too often consecutively, it seems the server stops to handle the request at all (or even receiving them). After about 3 or 4  BadRequestException/HeaderError groups (one per image), there's just silence in the log. When I then wait a little (a minute or so), I receive the errors on the log again.
Thinking of it, I also have the impression, that once on image fails to synchronize, the other synchronizations are stopped too, is that correct?

Concerning the byte size, I can only guess, as the problem occurs only for multiple images and the synchronization does not always fail at the same image. But I just now got two Exception messages on the server, where the expected byte size was the same, and the number of bytes actually written differs. Notably, the images I tried to attach are all of different sizes. So I guess the number of bytes written to the server actually differs for the same images, which seems kind of random to me.

**Comment by mhoeher** [comment]
Nope, unfortunately, the fix for opening attachments won't help for the mobile systems (the mechanism to open files is completely different comparing desktop (and there even sandboxed approaches like snap, flatpak, etc) and mobile systems. I have to do some research to find out how to properly implement opening of attachments on mobiles later on.

Regarding the sync: That sounds really strange and yes: random. The only idea that comes to my mind would be to reduce the number of parallel connections that the client opens at once. Maybe I could make this an option in the app. I'll play a little with this this evening and try to come up with a demo version - maybe this helps to at least mitigate the issue.

**Comment by mhoeher** [comment]
Hey, another short update: I just released another version of the app in the edge channel on Snapcraft. This version uses only one worker for the sync. Would be cool if you could give this version a try and see if it works more reliably.

**Comment by vloexo** [comment]
Hello, thanks for the update. As far as I can tell, this seems to have fixed the synchronization errors completely (in my case at least). I tried to upload multiple images "at once", and the sync succeeded. No more server exceptions either.
Thank you very much for your efforts :)

**Comment by mhoeher** [comment]
Thanks a lot for testing! That's in fact good news :wink: 

I tried to make up my mind about how to properly handle this (in the quick fix version you tested, I simply forced the sync to use at most one connection to the server at a time). This seemed to have cured the issue for your setup, however, this definitely is not a good solution for everyone (as especially the initial sync when adding an existing  library to a device would than again take ages to finish).

So I tried to implement the following:

1. We run a "normal" sync with a maximum number of jobs connecting to the server in parallel. For most setups (and I guess even in your case when not uploading larger files - I hope) this should just work.
2. If we detect that the server unexpectedly closed the connection, we will run immediately a second sync, but with only one job running at a time and hence connecting to the server.

It would be great if you could test this again with your setup and see if you still can successfully "mass upload" several attachments at once. A build is currently under way and should be available via `snap` once [this CI job](https://gitlab.com/rpdev/opentodolist/-/jobs/1213552416) finished.

With this version, you should see some sync errors in the sync log in the app, but overall, the sync should succeed.

**Comment by vloexo** [comment]
> It would be great if you could test this again with your setup and see if you still can successfully "mass upload" several attachments at once. A build is currently under way and should be available via `snap` once [this CI job](https://gitlab.com/rpdev/opentodolist/-/jobs/1213552416) finished.
> 
> With this version, you should see some sync errors in the sync log in the app, but overall, the sync should succeed.

Thank you for providing a reasonable solution to this problem :) "Mass-uploading" seems to work now, just as you described. And yes, I too think that for smaller files (like the note files themselves) it should work as intended also with parallel synchronization.

Unfortunately, another problem occurs to me with this version of the software, but I reckon it's just a minor bug and probably easy to fix:
Clicking on an attachment does still open the respective file as expected BUT it opens twice. So depending on the file type and software installed, as well as if I already opened the file type, I either get two windows to choose a program to open the file (if there are multiple candidates installed), or just two instances of the respective program open (e.g. two windows showing the same picture).
So my guess would be, that due to some changes you made, the function that handles the opening of attachments is somehow called twice.

**Comment by mhoeher** [comment]
First of all: Thanks for testing and confirming! So this will be definitely fixed in the next release then :wink: 

Regarding opening of attachments: You are right, there was a bug in :open_mouth: What I find interesting: I once call `xdg-open` manually from within the snap and *also* (this is the actual bug) into Qt's own implementation. I wonder why it now opens twice (and before not at all)? This is a bit... well... strange. But anyway, I pushed just another fix and a new snap version should be available shortly. Would be great if you could check with that one. If it works, I'd go ahead and merge the changes.

**Comment by vloexo** [comment]
I just tried opening attachments with the new version, and it works as intended again. One click - one instance of whatever viewer software is opened. So, this seems to be fixed as well. Thank you very much :)
Do you happen to have a PayPal address, so I might give out a coffee? (or whatever floats your boat ;) )

**Comment by mhoeher** [comment]
Great, thanks a lot! Just merged the changes - it's anyway nearly time for the next release :wink: 

Regarding Paypal... the usual disclaimer first: This is a hobby app of mine and if my users are happy, this is pretty much enough for me. Plus, keep in mind I could do something stupid with each release, so some day due to a new bug the app might just accidentally delete some data of yours and then you'll hate me :wink: If this isn't enough to scare you of: There's a "Donation" link in the app sidebar, where you can find appropriate links. Big thanks already!

Some bookkeeping at the end: I'll also go ahead and close this issue, as I think the main issue (image attachments) should be working now. For the other issues I created and linked above, I'd like to ask you to subscribe to them on Gitlab, if you are interested in progress on them. You can log in at Gitlab with your Github account, so no additional "genuine" login required. Alternatively, you can also just keep a close look on the future release notes.

And as always: If you find anything else, whether it is something you'd like to see added to the app or where you think this is a bug, please don't hesitate and raise further issues :wink:
